File: src/competitve practice/leetcode-30day-challenge/week3/MinimumPathSum.java
class Solution {

    public int calculateMin(int[][] dp, int[][] grid, int row, int col) {
        if (row == -1 || col == -1) {
            return Integer.MAX_VALUE;
        }
        if (row == 0 && col == 0) {
            return grid[0][0];
        }
        if (dp[row][col] == 0) {
            int min1 = calculateMin(dp, grid, row - 1, col);
            int min2 = calculateMin(dp, grid, row, col - 1);
            int min = Math.min(min1, min2);
            dp[row][col] = grid[row][col] + min;
        }
        return dp[row][col];
    }

    public int minPathSum(int[][] grid) {
        int[][] dp = new int[grid.length + 1][grid[0].length + 1];

        return calculateMin(dp, grid, grid.length - 1, grid[0].length - 1);

    }
}

File: LeetCode/leetcode-30day-challenge/week3/C++/MinimumPathSum.cpp
#include <bits/stdc++.h>
using namespace std;

//Type: Dynamic Programming

//Approach: Create a dp table and intialize it with zeros.
//          Since all the numbers in the grid are non-negative,
//          the minimum sum path for a row/column would be unique,i.e., straight line from starting pt(0,0)
//          For any other interior cell, a cost would be the current cell cost
//          in addition to minimum of the cost to reach the element above or left to it.
//Complexity: O(M*N) time , O(M*N) space complexity

class Solution
{
public:
    int minPathSum(vector<vector<int>> &grid)
    {
        ios_base::sync_with_stdio(false);
        cin.tie(NULL);
        cout.tie(NULL);
        int m = grid.size();
        int n = grid[0].size();
        int dp[m + 1][n + 1];
        memset(dp, 0, sizeof(dp));
        // dp[0][0]=grid[0][0];
        for (int i = 1; i <= m; i++)
        {
            for (int j = 1; j <= n; j++)
            {
                if (i == 1)
                {
                    dp[i][j] = dp[i][j - 1] + grid[i - 1][j - 1];
                }
                else if (j == 1)
                {
                    dp[i][j] = dp[i - 1][j] + grid[i - 1][j - 1];
                }
                else
                {
                    dp[i][j] = min(dp[i - 1][j], dp[i][j - 1]) + grid[i - 1][j - 1];
                }
                //cout << dp[i][j] << " ";
            }
            //cout << endl;
        }
        return dp[m][n];
    }
};